src/main/java/com/uber/cadence/internal/sync/DeterministicRunnerImpl.java

```diff
@@ -256,7 +256,7 @@ public void runUntilAllBlocked() throws Throwable {
             }
           } else {
             long t = c.getBlockedUntil();
-            if (t > nextWakeUpTime) {
+            if (nextWakeUpTime == 0 || t < nextWakeUpTime) {
```

Comment:
I tried tinkering with this pull request. Figured the `t < nextWakeUpTime` condition might not make sense if `t == 0`, since `t == 0` doesn't seem to represent a lower time period, it just represents an unset time. Ie, I rewrote this line as

```
if (nextWakeUpTime == 0 || t != 0 && t < nextWakeUpTime) {
    nextWakeUpTime = t;
}
```

This mitigates all test failures except one. The `WorkflowTestingTest.testTimerCancellation` still fails, since the following code inside `updateTimers()` (I don't follow why this is written this way) causes the timer not to be cancelled. Wondering if you're missing some additional changes around here?

```
if (wakeUpTime == nextWakeUpTime && timerCancellationHandler != null) {
  return; // existing timer
}
```

----------

[1] Full test failures below

```
com.uber.cadence.internal.testing.WorkflowTestingTest > testTimerCancellation FAILED
    java.lang.AssertionError at WorkflowTestingTest.java:586

com.uber.cadence.internal.sync.DeterministicRunnerTest > testJoinTimeout FAILED
    junit.framework.AssertionFailedError at DeterministicRunnerTest.java:641

com.uber.cadence.workflow.WorkflowTest > testChildWorkflowExecutionPromiseHandler[TestService Sticky OFF] FAILED
    com.uber.cadence.client.WorkflowTimedOutException at WorkflowTest.java:2951

com.uber.cadence.workflow.WorkflowTest > testTimerFiringTimestampEarlierThanExpected[TestService Sticky OFF] FAILED
    java.lang.RuntimeException at WorkflowTest.java:5181

com.uber.cadence.workflow.WorkflowTest > testChildAsyncLambdaWorkflow[TestService Sticky OFF] FAILED
    com.uber.cadence.client.WorkflowFailureException at WorkflowTest.java:1713
        Caused by: com.uber.cadence.workflow.ChildWorkflowTimedOutException

com.uber.cadence.workflow.WorkflowTest > testChildWorkflowExecutionPromiseHandler[TestService Sticky ON] FAILED
    com.uber.cadence.client.WorkflowTimedOutException at WorkflowTest.java:2951

com.uber.cadence.workflow.WorkflowTest > testTimerFiringTimestampEarlierThanExpected[TestService Sticky ON] FAILED
    java.lang.RuntimeException at WorkflowTest.java:5181

com.uber.cadence.workflow.WorkflowTest > testChildAsyncLambdaWorkflow[TestService Sticky ON] FAILED
    com.uber.cadence.client.WorkflowFailureException at WorkflowTest.java:1713
        Caused by: com.uber.cadence.workflow.ChildWorkflowTimedOutException
```